File: 0547-number-of-provinces/0547-number-of-provinces.cs
public class Solution {
    public int FindCircleNum(int[][] isConnected) {
        int[] visited = new int[isConnected.Length];
        int count = 0;
        for(int i=0; i<isConnected.Length; i++)
        {
            if(visited[i] == 0)
            {
                count++;
                dfs(i,isConnected,visited);
            }
        }
        return count;
    }
    private void dfs(int current, int[][] isConnected, int[] visited)
    {
        visited[current] = 1;
        for(int i=0; i<isConnected[0].Length; i++)
        {
            if(visited[i] == 0 && isConnected[current][i] == 1)
            {
                dfs(i,isConnected,visited);
            }
        }
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        DS ds = new DS(n);
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<n; j++)
            {
                if(i == j) continue;
                if(isConnected[i][j] == 1) ds.union(i,j);
            }
        }
        int ans = 0;
        for(int i=0; i<n; i++)
        {
            if(ds.findParent(i) == i)
            {
                ans++;
            }
        }
        return ans;
    }
    class DS {
        int[] par;
        int[] size;
        DS(int n)
        {
            par = new int[n];
            size = new int[n];
            
            for(int i=0; i<n; i++)
            {
                par[i] = i;
                size[i] = 1;
            }
        }
        public int findParent(int u)
        {
            if(par[u] == u)
            {
                return u;
            }
            return par[u] = findParent(par[u]);
        }
        public void union(int u, int v)
        {
            int pU = findParent(u);
            int pV = findParent(v);
            if(pU == pV) { return; }
            if(size[pU] >= size[pV])
            {
                size[pU] += size[pV];
                par[pV] = pU;
            }
            else
            {
                size[pV] += size[pU];
                par[pU] = pV;
            }
        }
    }
}